sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs

```diff
@@ -268,5 +269,61 @@ protected async Task WaitForProgressAsync(List<StorageProgress> progressList, lo
             // TODO: #7077 - These are too flaky/noisy so I'm changing to Warn
             Assert.Warn("Progress notifications never completed!");
         }
+
+        protected void AssertSecondaryStorageFirstRetrySuccessful(string primaryHost, string secondaryHost, TestExceptionPolicy testExceptionPolicy)
+        {
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[0]);
+            Assert.AreEqual(secondaryHost, testExceptionPolicy.Hosts[1]);
+        }
+
+        protected void AssertSecondaryStorageSecondRetrySuccessful(string primaryHost, string secondaryHost, TestExceptionPolicy testExceptionPolicy)
+        {
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[0]);
+            Assert.AreEqual(secondaryHost, testExceptionPolicy.Hosts[1]);
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[2]);
+        }
+
+        protected void AssertSecondaryStorageThirdRetrySuccessful(string primaryHost, string secondaryHost, TestExceptionPolicy testExceptionPolicy)
+        {
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[0]);
+            Assert.AreEqual(secondaryHost, testExceptionPolicy.Hosts[1]);
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[2]);
+            Assert.AreEqual(secondaryHost, testExceptionPolicy.Hosts[3]);
+        }
+
+        protected void AssertSecondaryStorage404OnSecondary(string primaryHost, string secondaryHost, TestExceptionPolicy testExceptionPolicy)
+        {
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[0]);
+            Assert.AreEqual(secondaryHost, testExceptionPolicy.Hosts[1]);
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[2]);
+            Assert.AreEqual(primaryHost, testExceptionPolicy.Hosts[3]);
+        }
+        protected delegate Task<object> GetPerformer
+            (object service);
+        protected delegate bool HasResponse(object response);
+
+        protected async Task<object> EnsurePropagatedAsync(
+            GetPerformer getPerformer,
+            HasResponse responseChecker,
+            object service)
+        {
+            bool responseReceived = false;
+            object response = null;
+            int count = 0;
+            while (!responseReceived && count < 90) // max of 15 minutes to allow for propagation to secondary host
```

Comment:
Move magic values to TestConstants?  Or even just declare a const MaxSecondaryPropTimeout or similar in this file?